docs: Update drift docs for Specified and Discovered drift (#4790)
aws/karpenter: website/content/en/preview/concepts/disruption.md

{{% /alert %}}

### Drift

> #### Specified Drift  

Reviewer: This reads very strangely to me. Thoughts on just `Spec Drift`, since its about changed applied to `spec`?

@jonathan-innis any thoughts?

Reviewer 2: Agreed. I think that Spec drift makes more sense

Reviewer 3: I originally was against using Specified/Spec, but I didn't have a better alternative.

Spec Drift sounds like Drift of the object's spec field, like NodeClassSpec.
I could imagine reading this and being confused why it's called that when SGs, Subnets, and AMIs and NodeRequirements aren't included in this.

Author: How would y'all feel if we dropped the naming all together? Just say drift for the general drift, special drift, and no drift

Reviewer 3: Talked with @engedaam, we're thinking it might be better to just call out "Drift has special behaviors for certain fields, but otherwise generally works like [Specified Drift]"

Reviewer 2: > being confused why it's called that when SGs, Subnets, and AMIs and NodeRequirements aren't included in this

I'm on-board with this except for the fact that I would still want to call it "Spec Drift". We're referencing the spec of an object and diffing the hash of that spec  to determine if something has drifted. Like you said, I think saying something like, "This is how drift works generally for Spec Drift, but there are some exceptions that are Discovered and here's the reason why these are different."

> Specified Drift handles drift changes that occur on the owning NodePool/EC2NodeClass. For Specified drift, values in the owning NodePool/EC2NodeClass are reflected in the NodeClaim in the same way that they’re set. A NodeClaim will be detected as drifted if the values in the NodePool/EC2NodeClass do not match the values in the NodeClaim. Similar to the upstream `deployment.spec.template` relationship to pods, Karpenter will annotate the owning NodePool and EC2NodeClass with a hash of the NodeClaim to check for drift. Some special cases will be discovered either from Karpenter or through the CloudProvider interface, triggered by NodeClaim/Instance/NodePool/EC2NodeClass changes.

Reviewer 3: ```suggestion
Specified Drift handles drift changes that occur on the owning NodePool/EC2NodeClass. For Specified drift, values in the owning NodePool/EC2NodeClass are reflected in the NodeClaim in the same way that they’re set. A NodeClaim will be detected as drifted if the values in the NodePool/EC2NodeClass do not match the values in the NodeClaim. Similar to the upstream `deployment.spec.template` relationship to pods, Karpenter will annotate the owning NodePool and EC2NodeClass with a hash of the NodeClaimTemplate to check for drift. Some special cases will be discovered either from Karpenter or through the CloudProvider interface, triggered by NodeClaim/Instance/NodePool/EC2NodeClass changes.
```

#### Behavioral Fields
Behavioral Fields are treated as over-arching settings on the NodePool to dictate how Karpenter behaves. These fields don’t correspond to settings on the NodeClaim or instance. They’re set by the user to control Karpenter’s Provisioning and disruption logic. Since these don’t map to a desired state of NodeClaims, __behavioral fields are not considered for Drift__.

Read the [Drift Design](https://github.com/aws/karpenter-core/blob/main/designs/drift.md) for more.

##### NodePool

> | Fields                     | Specified | Discoverd | 

Reviewer 3: ```suggestion
| Fields                     | Specified | Discovered | 
```

{{% /alert %}}

### Drift

> Spec Drift handles drift changes that occur on the owning NodePool/EC2NodeClass. For Spec drift, values in the owning NodePool/EC2NodeClass are reflected in the NodeClaim in the same way that they’re set. A NodeClaim will be detected as drifted if the values in the NodePool/EC2NodeClass do not match the values in the NodeClaim. Similar to the upstream `deployment.spec.template` relationship to pods, Karpenter will annotate the owning NodePool and EC2NodeClass with a hash of the NodeClaimTemplate to check for drift. Some special cases will be discovered either from Karpenter or through the CloudProvider interface, triggered by NodeClaim/Instance/NodePool/EC2NodeClass changes.

Author: Yeah, I think that is cleaner  for customers

#### Special Cases on Drift

> The special case on drift drift can correspond to multiple values and must be handled differently. Drift on resolved felids can create cases where drift occurs without changes to CRDs, or where CRD changes do not result in drift. For example, if a NodeClaim has `node.kubernetes.io/instance-type: m5.large`, and requirements change from `node.kubernetes.io/instance-type In [m5.large]` to `node.kubernetes.io/instance-type In [m5.large, m5.2xlarge]`, the NodeClaim will not be drifted because its value is still compatible with the new requirements. Conversely, if a NodeClaim is using a NodeClaim image `ami: ami-abc`, but a new image is published, Karpenter's `EC2NodeClass.spec.amiSelectorTerms` will discover that the new correct value is `ami: ami-xyz`, and detect the NodeClaim as drifted.

Reviewer 3: ```suggestion
In special cases, drift can correspond to multiple values and must be handled differently. Drift on resolved fields can create cases where drift occurs without changes to CRDs, or where CRD changes do not result in drift. For example, if a NodeClaim has `node.kubernetes.io/instance-type: m5.large`, and requirements change from `node.kubernetes.io/instance-type In [m5.large]` to `node.kubernetes.io/instance-type In [m5.large, m5.2xlarge]`, the NodeClaim will not be drifted because its value is still compatible with the new requirements. Conversely, if a NodeClaim is using a NodeClaim image `ami: ami-abc`, but a new image is published, Karpenter's `EC2NodeClass.spec.amiSelectorTerms` will discover that the new correct value is `ami: ami-xyz`, and detect the NodeClaim as drifted.
```

{{% /alert %}}

### Drift

> Drift handles drift changes that occur on the owning NodePool/EC2NodeClass. For Drift, values in the owning NodePool/EC2NodeClass are reflected in the NodeClaimTemplateSpec/EC2NodeClassSpec in the same way that they’re set. A NodeClaim will be detected as drifted if the values in the NodePool/EC2NodeClass do not match the values in the NodeClaim. Similar to the upstream `deployment.spec.template` relationship to pods, Karpenter will annotate the owning NodePool and EC2NodeClass with a hash of the NodeClaimTemplateSpec to check for drift. Some special cases will be discovered either from Karpenter or through the CloudProvider interface, triggered by NodeClaim/Instance/NodePool/EC2NodeClass changes.

Reviewer 3: thinking we can trim it down even more, thoughts?
```suggestion
Drift handles changes to the NodePool/EC2NodeClass. For Drift, values in the NodePool/EC2NodeClass are reflected in the NodeClaimTemplateSpec/EC2NodeClassSpec in the same way that they’re set. A NodeClaim will be detected as drifted if the values in its owning NodePool/EC2NodeClass do not match the values in the NodeClaim. Similar to the upstream `deployment.spec.template` relationship to pods, Karpenter will annotate the owning NodePool and EC2NodeClass with a hash of the NodeClaimTemplateSpec to check for drift. Some special cases will be discovered either from Karpenter or through the CloudProvider interface, triggered by NodeClaim/Instance/NodePool/EC2NodeClass changes.
```

## Drift

### Specification Drift  
Specification Drift on most fields are only triggered by changes to the owning CustomResource. Some special cases will be reconciled discovery, triggered by Machine/Node/Instance changes or Provisioner/AWSNodeTemplate changes. For specification drift , values in the CustomResource are reflected in the Machine in the same way that they’re set. A machine will be detected as drifted if the values in the CRDs do not match the values in the Machine. By default, fields are drifted using specification drift. 

> ### Discovery Drift

Reviewer: What do you think about Discovered?
